File: RLBridge/communication.lua
--- RLBridge communication module
--- Handles line-delimited JSON communication with the external AI system over a
--- localhost TCP socket. The Python trainer is the server; the game is the client.

local COMM = {}
local utils = require("utils")
local json = require("dkjson")

-- LuaSocket ships inside LÖVE; fall back to the raw core module if the
-- pure-Lua wrapper is unavailable
local ok, socket = pcall(require, "socket")
if not ok then
    ok, socket = pcall(require, "socket.core")
end
if not ok then
    socket = nil
end

-- Socket communication settings with a persistent connection
local comm_enabled = false
local host = os.getenv("BALATRO_RL_HOST") or "127.0.0.1"
local port = tonumber(os.getenv("BALATRO_RL_PORT") or "12345")
local conn = nil
local last_connect_attempt = 0
local RECONNECT_COOLDOWN = 1 -- seconds between connect attempts

--- Initialize socket communication
--- Sets up a persistent connection with the external AI system
--- @return nil
function COMM.init()
    utils.log_comm("Initializing socket communication...")
    comm_enabled = true -- Enable communication, connection opens on first use
end

--- Lazy initialization of the socket connection when first needed
--- @return boolean True if the connection is ready, false otherwise
function COMM.ensure_connected()
    if conn then
        return true -- Already connected
    end

    if not socket then
        utils.log_comm("ERROR: LuaSocket unavailable, cannot connect")
        return false
    end

    -- AI.update runs every frame; throttle retries so a missing trainer
    -- does not stall the game with a connect attempt per frame
    local now = os.time()
    if now - last_connect_attempt < RECONNECT_COOLDOWN then
        return false
    end
    last_connect_attempt = now

    local sock, err = socket.tcp()
    if not sock then
        utils.log_comm("ERROR: Cannot create socket: " .. tostring(err))
        return false
    end

    -- Bounded connect so a missing trainer does not hang the game forever
    sock:settimeout(2)
    local connected, connect_err = sock:connect(host, port)
    if not connected then
        utils.log_comm("ERROR: Cannot connect to AI at " .. host .. ":" .. port ..
            " (" .. tostring(connect_err) .. ")")
        sock:close()
        return false
    end

    -- Block indefinitely on reads; the AI decides when it is done thinking
    sock:settimeout(nil)
    pcall(function() sock:setoption("tcp-nodelay", true) end)
    conn = sock

    utils.log_comm("Connected to AI at " .. host .. ":" .. port)
    return true
end

--- Send game turn request to AI and get action via the persistent socket
--- @param game_state table Current game state data
--- @param available_actions table Available actions list
--- @param timing table|nil Game-side latency counters since the last action
--- @return table|nil Action response from AI, nil if error
function COMM.request_action(game_state, available_actions, timing)
    if not comm_enabled then
        utils.log_comm("ERROR: Communication not enabled")
        return nil
    end

    -- Lazy initialization - connect when first needed
    if not COMM.ensure_connected() then
        utils.log_comm("ERROR: Failed to establish connection")
        return nil
    end

    local request = {
        game_state = game_state,
        available_actions = available_actions or {},
        timing = timing or {},
    }

    utils.log_comm(utils.get_timestamp() .. " Sending action request for state: " ..
        tostring(game_state.state) .. " (" .. utils.get_state_name(game_state.state) .. ")")

    -- Encode request as JSON
    local json_data = json.encode(request)
    if not json_data then
        utils.log_comm("ERROR: Failed to encode request JSON")
        return nil
    end

    -- Write request to the socket
    local sent, send_err = conn:send(json_data .. "\n")
    if not sent then
        utils.log_comm("ERROR: Failed to send request: " .. tostring(send_err))
        COMM.drop_connection()
        return nil
    end

    -- Read response from the socket
    local response_json, recv_err = conn:receive("*l")

    if not response_json or response_json == "" then
        utils.log_comm("ERROR: No response received from AI: " .. tostring(recv_err))
        COMM.drop_connection()
        return nil
    end

    local response_data = json.decode(response_json)
    if not response_data then
        utils.log_comm("ERROR: Failed to decode response JSON")
        return nil
    end

    utils.log_comm(utils.get_timestamp() .. " AI action: " .. tostring(response_data.action))
    return response_data
end

--- Drop a dead connection so the next request reconnects
--- @return nil
function COMM.drop_connection()
    if conn then
        conn:close()
        conn = nil
    end
end

--- Check if socket communication is enabled
--- Returns the current communication status
--- @return boolean True if enabled, false otherwise
function COMM.is_connected()
    return comm_enabled
end

--- Close communication
--- Terminates the persistent connection with the AI system
--- @return nil
function COMM.close()
    comm_enabled = false
    COMM.drop_connection()
    utils.log_comm("Socket communication closed")
end

return COMM

File: ai/utils/communication.py
"""
Balatro Socket Communication Abstraction

Pure line-delimited JSON I/O layer over a localhost TCP socket for communicating
with the Balatro mod. Handles low-level transport without any game logic or AI logic.

Python acts as the server (it owns the training lifecycle); the Balatro mod connects
as a client and reconnects on game relaunch.

This abstraction can be used by:
- balatro_env.py (for SB3 training)
- file_watcher.py (for testing)
- Any other component that needs to talk to Balatro
"""

import json
import logging
import os
import socket
from typing import Dict, Any, Optional

from ..telemetry import EventType, Stopwatch, emit

DEFAULT_HOST = os.environ.get("BALATRO_RL_HOST", "127.0.0.1")
DEFAULT_PORT = int(os.environ.get("BALATRO_RL_PORT", "12345"))


class BalatroSocketIO:
    """
    Clean abstraction for socket communication with the Balatro mod

    Responsibilities:
    - Accept a connection from the Balatro mod
    - Read JSON requests from the mod
    - Write JSON responses back to the mod
    - Keep the connection open persistently across an episode
    - Re-accept transparently when Balatro restarts
    """

    def __init__(self, host: str = DEFAULT_HOST, port: int = DEFAULT_PORT):
        self.host = host
        self.port = port
        self.logger = logging.getLogger(__name__)

        self.server: Optional[socket.socket] = None
        self.client: Optional[socket.socket] = None
        self.reader = None
        self.writer = None

        # Latency instrumentation. t_wait is time blocked on the game and is
        # expected to dominate the step budget; t_send is our own write cost.
        self.last_send_s = 0.0
        self.last_wait_s = 0.0
        self.round_trips = 0
        self._wait_watch = Stopwatch()

        self.start_server()
        self.accept_connection()

    def start_server(self) -> None:
        """
        Bind and listen on the configured host/port

        Safe to call once per instance; raises if the port is already taken.
        """
        try:
            self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.server.bind((self.host, self.port))
            self.server.listen(1)
            self.logger.info(f"Listening on {self.host}:{self.port}")
        except Exception as e:
            self.logger.error(f"Failed to bind {self.host}:{self.port}: {e}")
            self.cleanup()
            raise RuntimeError(f"Could not start communication server: {e}")

    def accept_connection(self, timeout: Optional[float] = None) -> None:
        """
        Wait for the Balatro mod to connect

        Replaces any previous connection. Disables Nagle so the
        request/response ping-pong is not delayed.

        Args:
            timeout: Seconds to wait, or None to block indefinitely. A bounded
                wait is what lets a mid-run disconnect be retried rather than
                hanging forever on a game that is never coming back.

        Raises:
            TimeoutError: If no connection arrived within the timeout
            RuntimeError: If accepting failed for any other reason
        """
        self.close_client()

        try:
            self.logger.info("🔧 Waiting for Balatro to connect...")
            self.logger.info("   Press 'R' in Balatro now to activate RL training!")

            self.server.settimeout(timeout)
            try:
                self.client, addr = self.server.accept()
            finally:
                # Leave the listening socket blocking for any other caller
                self.server.settimeout(None)
            self.client.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            self.reader = self.client.makefile("r", encoding="utf-8", newline="\n")
            self.writer = self.client.makefile("w", encoding="utf-8", newline="\n")

            self.logger.info(f"Balatro connected from {addr[0]}:{addr[1]}")
            emit(EventType.COMM, event="connected", peer=f"{addr[0]}:{addr[1]}")
        except socket.timeout:
            # Distinct from a real failure: the caller may want to retry
            self.close_client()
            raise TimeoutError(f"No Balatro connection within {timeout}s")
        except Exception as e:
            self.logger.error(f"Failed to accept connection: {e}")
            self.close_client()
            raise RuntimeError(f"Could not accept Balatro connection: {e}")

    def ensure_connected(self, timeout: Optional[float] = None) -> bool:
        """
        Re-accept a connection if the previous one was dropped

        Args:
            timeout: Seconds to wait for a new connection, None to block

        Returns:
            True if a live connection is available
        """
        if self.client:
            return True

        try:
            self.accept_connection(timeout=timeout)
            return True
        except Exception:
            return False

    def wait_for_request(self, accept_timeout: Optional[float] = None) -> Optional[Dict[str, Any]]:
        """
        Wait for a new request from the Balatro mod

        Blocks until Balatro writes a request to the socket.

        Args:
            accept_timeout: If a reconnection is needed, how long to wait for it

        Returns:
            Parsed JSON request data, or None on disconnect/error
        """
        if not self.ensure_connected(timeout=accept_timeout):
            return None

        try:
            # Everything between our response and the game's next request is
            # game-side time: animations, event queue, frames where the state
            # hash did not change. This is the number that matters.
            self._wait_watch.start()
            request_line = self.reader.readline().strip()
            self.last_wait_s = self._wait_watch.stop()

            if not request_line:
                # Empty read means the peer closed the connection
                self.logger.warning("Balatro disconnected")
                emit(EventType.COMM, event="disconnected")
                self.close_client()
                return None

            request_data = json.loads(request_line)
            self.round_trips += 1
            self.logger.debug(f"📥 RECEIVED REQUEST: {request_line}")
            return request_data

        except json.JSONDecodeError as e:
            self.logger.error(f"Invalid JSON in request: {e}")
            emit(EventType.COMM, event="bad_json", error=str(e))
            return None
        except Exception as e:
            self.logger.error(f"Error reading from socket: {e}")
            emit(EventType.COMM, event="read_error", error=str(e))
            self.close_client()
            return None

    def send_response(self, response_data: Dict[str, Any]) -> bool:
        """
        Send a response back to the Balatro mod

        Args:
            response_data: Response dictionary to send

        Returns:
            True if successful, False if error
        """
        if not self.ensure_connected():
            self.logger.error("No connection to Balatro")
            return False

        watch = Stopwatch().start()
        try:
            json.dump(response_data, self.writer)
            self.writer.write("\n")  # Important: newline delimits the message
            self.writer.flush()  # Force write to socket immediately
            self.last_send_s = watch.stop()

            self.logger.debug(f"📤 SENT RESPONSE: {json.dumps(response_data)}\n\n")
            return True

        except Exception as e:
            self.last_send_s = watch.stop()
            self.logger.error(f"❌ ERROR sending response: {e}")
            import traceback
            self.logger.error(f"❌ TRACEBACK: {traceback.format_exc()}")
            emit(EventType.COMM, event="send_error", error=str(e))
            self.close_client()
            return False

    def timings(self) -> Dict[str, float]:
        """
        Latency of the most recent round trip

        Returns:
            Dict with t_send (our write cost) and t_wait (blocked on the game)
        """
        return {"t_send": self.last_send_s, "t_wait": self.last_wait_s}

    def close_client(self):
        """
        Close the current client connection and its stream wrappers
        """
        for stream in (self.reader, self.writer):
            if stream:
                try:
                    stream.close()
                except Exception as e:
                    self.logger.debug(f"Failed to close stream: {e}")
        self.reader = None
        self.writer = None

        if self.client:
            try:
                self.client.close()
                self.logger.debug("Closed client socket")
            except Exception as e:
                self.logger.warning(f"Failed to close client socket: {e}")
            self.client = None

    def cleanup(self):
        """
        Close the client connection and release the listening port
        """
        self.close_client()

        if self.server:
            try:
                self.server.close()
                self.logger.debug("Closed server socket")
            except Exception as e:
                self.logger.warning(f"Failed to close server socket: {e}")
            self.server = None

        self.logger.debug("Socket communication cleanup complete")


# Backwards-compatible alias for the previous named-pipe implementation
BalatroPipeIO = BalatroSocketIO
